engine/src/main/java/org/hibernate/validator/internal/engine/valueextraction/ValueExtractorResolver.java

```diff
@@ -303,26 +303,27 @@ else if ( TypeHelper.isAssignable( descriptor.getContainerType(), candidate.getC
 		}
 
 		Set<ValueExtractorDescriptor> valueExtractorDescriptors = possibleValueExtractorsByRuntimeType.get( runtimeType );
-		if ( valueExtractorDescriptors == null ) {
-			//otherwise we just look for maximally specific extractors for the runtime type
-			Set<ValueExtractorDescriptor> possibleValueExtractors = potentialValueExtractorDescriptors
-					.stream()
-					.filter( e -> TypeHelper.isAssignable( e.getContainerType(), runtimeType ) )
-					.collect( Collectors.toSet() );
-
-			valueExtractorDescriptors = getMaximallySpecificValueExtractors( possibleValueExtractors );
+
+		if ( valueExtractorDescriptors != null ) {
+			return valueExtractorDescriptors;
 		}
 
+		Set<ValueExtractorDescriptor> possibleValueExtractors = potentialValueExtractorDescriptors
+				.stream()
+				.filter( e -> TypeHelper.isAssignable( e.getContainerType(), runtimeType ) )
+				.collect( Collectors.toSet() );
+
+		valueExtractorDescriptors = getMaximallySpecificValueExtractors( possibleValueExtractors );
+
 		if ( valueExtractorDescriptors.isEmpty() ) {
 			nonContainerTypes.put( runtimeType, NON_CONTAINER_VALUE );
```

Comment:
@gsmet, not directly related to this change, but can you check the beginning of the two methods:

```
if ( nonContainerTypes.contains( runtimeType ) ) {
```

Shouldn't this be `containsKey()` instead? It's not wrong per se, but it currently defeats the purpose of the `nonContainerTypes` cache.

Another question is the usage of the unbounded maps; in `BeanMetaDataManagerImpl` we're using `ConcurrentReferenceHashMap` instead, which would release the entries if there's need.

Reply:
Fixed the lookup issue by using a Set, thanks for the `newSetFromMap()` trick.

As discussed, all these elements are properly cleared when cleared when closing the VF so we are safe on this side.